packages/ajax/test/AjaxClass.interceptors-cache.test.js

```diff
@@ -0,0 +1,713 @@
+      })
+      .then(() => ajax.get('/foo-request-2'))
+      .then(() => {
+        expect(server.requests.length).to.equal(3);
+      })
+      .then(() => ajax.post('/test', {}, actionConfig))
+      .then(() => {
+        expect(server.requests.length).to.equal(4);
+      })
+      .then(() => ajax.get('/foo-request-1'))
+      .then(() => {
+        expect(server.requests.length).to.equal(5);
+      })
+      .then(() => ajax.get('/foo-request-2'))
+      .then(() => {
+        expect(server.requests.length).to.equal(5);
+      })
+      .finally(() => {
+        removeCacheInterceptors(ajax, indexes);
+      });
+  });
+
+  it('previously cached data has to be invalidated when regex invalidation rule triggered and urls are nested', () => {
+    newCacheId();
+
+    const indexes = addCacheInterceptors(ajax, {
+      useCache: 'always',
+      timeToLive: 1000,
+    });
+
+    server.respondWith(/^\/posts$/, xhr => {
+      xhr.respond(200, { 'Content-Type': 'application/json' }, '{}');
+    });
+
+    server.respondWith(/^\/posts\/1$/, xhr => {
+      xhr.respond(200, { 'Content-Type': 'application/json' }, '{}');
+    });
+
+    const actionConfig = {
+      lionCacheOptions: {
+        invalidateUrlsRegex: /posts/gi,
+      },
+    };
+
+    return ajax
+      .get('/test', actionConfig)
+      .then(() => ajax.get('/test'))
+      .then(() => {
+        expect(server.requests.length).to.equal(1);
+      })
+      .then(() => ajax.get('/posts'))
+      .then(() => {
+        expect(server.requests.length).to.equal(2);
+      })
+      .then(() => ajax.get('/posts'))
+      .then(() => {
+        expect(server.requests.length).to.equal(2);
+      })
+      .then(() => ajax.get('/posts/1'))
+      .then(() => {
+        expect(server.requests.length).to.equal(3);
+      })
+      .then(() => ajax.get('/posts/1'))
+      .then(() => {
+        expect(server.requests.length).to.equal(3);
+      })
+      .then(() => ajax.post('/test', {}, actionConfig))
+      .then(() => {
+        expect(server.requests.length).to.equal(4);
+      })
+      .then(() => ajax.get('/posts'))
+      .then(() => {
+        expect(server.requests.length).to.equal(5);
+      })
+      .then(() => ajax.get('/posts/1'))
+      .then(() => {
+        expect(server.requests.length).to.equal(5);
+      })
+      .finally(() => {
+        removeCacheInterceptors(ajax, indexes);
+      });
+  });
+
+  it('does not returns the cached object on second call default timeToLive`', () => {
+    newCacheId();
+
+    const indexes = addCacheInterceptors(ajax, {
+      useCache: 'always',
+      timeToLive: 0,
+    });
+
+    const ajaxGetSpy = sinon.spy(ajax, 'get');
+
+    return ajax
+      .get('/test')
+      .then(() => {
+        const clock = sinon.useFakeTimers();
+        expect(ajaxGetSpy.calledOnce).to.be.true;
+        expect(ajaxGetSpy.calledWith('/test')).to.be.true;
+        clock.tick(1);
+        clock.restore();
+      })
```

Comment:
I don't really understand what this tests? We're spying ajax.get and we call it with '/test', so it seems quite obvious to me that the two expects would be be true, even without a fake timer? Is the fake timer only used so that the next request happens 1 ms later and gets a normal response because it is past the 0ms timeToLive expiration? if that's the case, shouldn't we restore the clock in line 378 after the expect?

Reply:
ah, the reason is `clock` for some reason does not work in asynchronous  function, I was checking for an example in lion-web but could not find, but this one works like this